dvc/parsing/__init__.py

```diff
@@ -399,7 +399,7 @@ def normalized_iterable(self):
         """Convert sequence to Mapping with keys normalized."""
         iterable = self.resolved_iterable
         if isinstance(iterable, Mapping):
-            return iterable
+            return {str(k): v for k, v in iterable.items()}
```

Comment:
One issue I see with this fix is that in some cases keys will be interpreted as float, e.g: 
```yaml
stages:
  build:
    foreach:
      .0:
        epochs: 3
        thresh: 10
    do:
      ...
```

In this case `.0` will be interpreted as `0.0` (float) and thus the resolved name for the stage will be `build@0.0` even though it is defined as `.0` in the yaml. I'm not sure if we want to try and fix it for that case, since it depends on how we load yaml files in general.

Reply:
The thing that matters more than the representation is that the keys be unique.
In YAML, mapping keys are unique. So, parsing the following YAML either fails or overrides:
```yaml
0.0: 1
.0: 2
```

```py
>>> yaml.load(open(file))
{0.0: 2, True: 2, 'true': 3}
>>> load_yaml(file)  # ruamel.yaml fails
```

This however may break, however when the keys are mixed with stringy and non-stringy values (eg: (true, "true"), (0.0, "0.0"), etc.), which is already the case.
However, if we override/ignore either of the keys consistently, that's okay.

Reply:
It seems like it might be better for us to just require that `foreach` dictionary keys be strings since it removes any potential confusion over duplicate or ambiguous keys that may occur due to yaml int/float parsing?

So in the original example the user would need to quote the numeric keys so they are loaded as strings
```
stages:
  build:
    foreach:
      "2021":
        epochs: 3
        thresh: 10
      "2022":
        epochs: 10
        thresh: 15
```

Reply:
It's okay to support any possible values in yaml. It may be more confusing if we don't. From https://github.com/iterative/dvc/issues/8903, it does seem useful.

Reply:
For me, as the creator of the issue, it would be fine if either it was described in the documentation that numbers need quotation marks or the feedback comes from DVC that only strings are okay or that the number needs quotation marks or both. I have also solved that way.

Reply:
> It's okay to support any possible values in yaml.

I'm not sure it's possible for us to support this since we have to be able to address the foreach items as with strings.

```yaml
stages:
  build:
    foreach:
      "2021":
        epochs: 3
        thresh: 10
      2021:
        epochs: 10
        thresh: 15
```

This is legal yaml, and contains two unique foreach keys, but both stages would be addressed as `build@2021`.

Reply:
it seems to be stuck a bit, I think we need to pick this up, finish and merge

My 2 cents here - restrict non-strings is preferable - KISS - more explicit and easy to understand behavior, trying to support numbers and dealing with implicit precedence behavior will just be confusing and more complex to support imo. Suggest to avoid unless it's needed for some very good reason.

Reply:
We discussed this at our last meeting and decided to support them. 

This is part of parametrization/templating and interpolating/interpreting the variables is what it does.

Supporting it is more in line with KISS than restricting it imo. The confusing behaviour only happens if the two different values get interpreted in the same way, which is rare. Also, since it is a dictionary, only one of them will be used by dvc is expected by the user.

Reply:
> Supporting it is more in line with KISS than restricting it imo

Disagree in this case, but commit 😃 
Thanks for picking it up, this was stale for too long